### PR TITLE
Index Manager handles index states

### PR DESCRIPTION
**Pull request by wu-hui**

### CODE REVIEW

File: Firestore/core/src/local/memory_remote_document_cache.h [hunk 1]
```diff
@@ -64,6 +66,8 @@ class MemoryRemoteDocumentCache : public RemoteDocumentCache {
 
   // This instance is owned by MemoryPersistence; avoid a retain cycle.
   MemoryPersistence* persistence_;
+  // This instance is also owned by MemoryPersistence.
+  MemoryIndexManager* index_manager_;
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/memory_remote_document_cache.h:70]: Also `nullptr`.

File: Firestore/core/src/model/field_index.h [hunk 2]
```diff
@@ -114,11 +126,20 @@ class IndexOffset : public util::Comparable<IndexOffset> {
...
+  model::BatchId largest_batch_id() const {
+    return largest_batch_id_;
+  }
+
   util::ComparisonResult CompareTo(const IndexOffset& rhs) const;
 
  private:
   SnapshotVersion read_time_;
   DocumentKey document_key_;
+  model::BatchId largest_batch_id_ = InitialLargestBatchId();
```
Comment (schmidt-sebastian) [inline Firestore/core/src/model/field_index.h:142]: This is always set in the constructor.

**Review by schmidt-sebastian — APPROVED** [review]

File: Firestore/core/src/local/leveldb_index_manager.cc [hunk 3]
```diff
@@ -335,13 +352,29 @@ LevelDbIndexManager::GetDocumentsMatchingTarget(model::FieldIndex field_index,
...
-  (void)offset;
+  HARD_ASSERT(started_, "IndexManager not started");
+
+  ++memoized_max_sequence_number_;
+  for (const auto& field_index : GetFieldIndexes(collection_group)) {
+    IndexState updated_state{memoized_max_sequence_number_, offset};
+
+    auto state_key = LevelDbIndexStateKey::Key(uid_, field_index.index_id());
+    auto val = EncodeIndexState(updated_state);
+    db_->current_transaction()->Put(state_key, EncodeIndexState(updated_state));
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/local/leveldb_index_manager.cc]: Done.

File: Firestore/core/src/local/leveldb_persistence.cc [hunk 4]
```diff
@@ -253,7 +256,7 @@ LevelDbBundleCache* LevelDbPersistence::bundle_cache() {
   return bundle_cache_.get();
 }
 
-LevelDbDocumentOverlayCache* LevelDbPersistence::document_overlay_cache(
+LevelDbDocumentOverlayCache* LevelDbPersistence::GetDocumentOverlayCacheForUser(
     const User& user) {
   users_.insert(user.uid());
   current_document_overlay_cache_ =
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/local/leveldb_persistence.cc:262]: I suspect whoever made it this way thought keep C++ instances around was not a great idea.

File: Firestore/core/src/local/leveldb_persistence.h [hunk 5]
```diff
@@ -81,7 +81,7 @@ class LevelDbPersistence : public Persistence {
 
   LevelDbBundleCache* bundle_cache() override;
 
-  LevelDbDocumentOverlayCache* document_overlay_cache(
+  LevelDbDocumentOverlayCache* GetDocumentOverlayCacheForUser(
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/local/leveldb_persistence.h]: Done.

**Review by wu-hui — COMMENTED** [review]
There is a bug in the code earlier, basically the ownership of index manager in local store was wrong. It was creating (and deleting) multiple instances of index managers. 

So the new commits are more than just addressing feedback.

File: Firestore/core/src/local/leveldb_mutation_queue.cc [hunk 6]
```diff
@@ -126,9 +126,10 @@ BatchId LoadNextBatchIdFromDb(DB* db) {
 
 LevelDbMutationQueue::LevelDbMutationQueue(const User& user,
                                            LevelDbPersistence* db,
+                                           LevelDbIndexManager* index_manager,
```
Comment (schmidt-sebastian) [inline Firestore/core/src/local/leveldb_mutation_queue.cc:129]: Do we need LevelDbIndexManager here? We can remove the cast if we just pass in IndexManager.

**Review by schmidt-sebastian — APPROVED** [review]

File: Firestore/core/src/local/leveldb_mutation_queue.cc (hunk 6 above)
Comment (wu-hui) [reply Firestore/core/src/local/leveldb_mutation_queue.cc]: Done.

**Pull request merged**